### PR TITLE
Add AffineVectoreReduce instruction to the pxa layer.

### PR DESCRIPTION
**Pull request by LLITCHEV**
This is needed to allow us to use the data flow analysis mechanisms that the AffineReduceOp exposes/utilizes, so Fusion and other passes can operate on the AffineVectorReduceOp.

### CODE REVIEW

File: pmlc/dialect/pxa/tests/ops.mlir [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// RUN: pmlc-opt -pxa-dataflow-opt -canonicalize -convert-pxa-to-affine %s | FileCheck %s
+
+// CHECK-LABEL: func @pxa_vector_reduce
+func @pxa_vector_reduce(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>, %arg2: memref<100x100xf32>) {
```
Comment (jbruestle) [inline pmlc/dialect/pxa/tests/ops.mlir:4]: At the PXA level, outputs are typically returned rather than written into, i.e:

func @pxa_vector_reduce(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>)

Then, inside the body, allocate the output (with alloc), add an affine.parallel that clears it via pxa.reduce assign, then do the vector reductions and then return the result of the vector reduction.

Something like

%a = alloc<...>
%b = affine.parallel ... {
  %cst = constant.zero
  %c = pxa.reduce assign %cst, %b[...] ...
  yield %c
}
%d = affine.parallel (existing AP) {
  ...
  %e = pxa.vector_reduce add %vec, %b : ...
  yield %e
}
return %d

Look at the EDSL tests IR for more complete examples, it's a bit weird.  This should lower to very similar code to what you already lower to.

Basically, the current test doesn't actually use the PXA memref-state semantics (and had you added a CSE pass, the vector_reduce would have vanished I think since it's NoSideEffect and it's return is unused)

File: pmlc/dialect/pxa/ir/ops.td [hunk 2]
```diff
@@ -29,8 +29,7 @@ class PXA_OpWithPP<string mnemonic, list<OpTrait> traits = []> :
 
 def AnyStdScalar : AnyTypeOf<[AnyFloat, AnySignlessInteger]> {}
 
-// TODO: Add NoSideEffect
-def AffineReduceOp : PXA_OpWithPP<"reduce", []> {
+def AffineReduceOp : PXA_OpWithPP<"reduce", [NoSideEffect]> {
```
[the remark this answers is not in the document]
Comment (LLITCHEV) [reply pmlc/dialect/pxa/ir/ops.td:32]: I just saw the TODO: and added the "no side effect" flag, but I can add them back and just deal with it later.

[the remark this answers is not in the document]
Comment (flaub) [reply pmlc/dialect/pxa/ir/ops.td:32]: IIRC, the `Arg<AnyMemRef, "the memref to reduce", [MemRead, MemWrite]>:$mem,` supercedes `[NoSideEffect]` as far as the canonicalizer goes. I believe this is why I needed to write a separate canoncalizer pattern.

File: pmlc/dialect/pxa/ir/ops.cc [hunk 3]
```diff
@@ -126,6 +136,22 @@ struct SimplifyDeadReduce : public OpRewritePattern<AffineReduceOp> {
   }
 };
 
+/// Fold vector_reduce operations with no uses. Vector_reduce has side effects
+/// on the heap, but can still be deleted if it has zero uses.
+struct SimplifyDeadVectorReduce
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.cc:141]: I think we can reuse the existing `SimplifyDeadReduce`, the only difference is the template parameter.

File: pmlc/dialect/pxa/tests/ops.mlir (hunk 1 above)
Comment (LLITCHEV) [reply pmlc/dialect/pxa/tests/ops.mlir]: Added more tests. and returning the value now.

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc [hunk 4]
```diff
@@ -224,6 +227,88 @@ struct AffineReduceOpConversion : public LoweringBase<pxa::AffineReduceOp> {
...
+  }
+
+  Value createVectorReduction(ConversionPatternRewriter &rewriter,
+                              pxa::AffineVectorReduceOp op,
+                              Value source) const {
+    // Must be a VectorType
+    assert(source.getType().isa<VectorType>());
+
+    switch (op.agg()) {
+    case AggregationKind::assign:
```
[the remark this answers is not in the document]
Comment (LLITCHEV) [reply pmlc/conversion/pxa_to_affine/pxa_to_affine.cc:253]: Added more tests.

File: pmlc/dialect/pxa/ir/ops.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (LLITCHEV) [reply pmlc/dialect/pxa/ir/ops.td]: NoSideEffect flag removed (as it was before).

File: pmlc/dialect/pxa/ir/ops.cc [hunk 5]
```diff
@@ -126,6 +136,22 @@ struct SimplifyDeadReduce : public OpRewritePattern<AffineReduceOp> {
...
+    : public OpRewritePattern<AffineVectorReduceOp> {
```
[the remark this answers is not in the document]
Comment (LLITCHEV) [reply pmlc/dialect/pxa/ir/ops.cc:142]: NoSideEffect flag removed (as it was before).

File: pmlc/dialect/pxa/ir/ops.cc (hunk 3 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/ir/ops.cc:141]: I agree, you should template SimplyDeadReduce on the operation type to and then just instantiate two versions in the patterns to reduce code duplication.

File: pmlc/dialect/pxa/tests/ops.mlir [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+// CHECK-LABEL: func @pxa_reduce_mul
+func @pxa_reduce_mul(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>) {
+  %a = alloc() : memref<100x100xf32>
+  %r = affine.parallel (%i, %j, %k) = (0, 0, 0) to (100, 100, 100) reduce ("assign") -> (memref<100x100xf32>) {
+    %0 = affine.load %arg1[%i, %k] : memref<100x100xf32>
+    %1 = affine.load %arg0[%k, %j] : memref<100x100xf32>
+    %2 = mulf %0, %1 : f32
+    %red = pxa.reduce mul %2, %a[%i, %j] :  memref<100x100xf32>
+    // CHECK: %{{.*}} = affine.load %{{.*}} : memref<100x100xf32>
```
Comment (jbruestle) [inline pmlc/dialect/pxa/tests/ops.mlir:73]: The patterns are a bit needlessly permissive. You should probably do captures of the values.  You could verify that for example that the loaded value is passed into the mulf and that the output of the mulf goes into the store, and that the store and the load both reference the same buffer (which they should post pxa->affine)

File: pmlc/dialect/pxa/ir/ops.cc (hunk 3 above)
Comment (LLITCHEV) [reply pmlc/dialect/pxa/ir/ops.cc]: It is done.

File: pmlc/dialect/pxa/tests/ops.mlir (hunk 6 above)
Comment (LLITCHEV) [reply pmlc/dialect/pxa/tests/ops.mlir]: Added more soecific checks.

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc [hunk 7]
```diff
@@ -236,9 +237,16 @@ struct AffineVectorReduceOpConversion
...
+    auto srcMapAttr = AffineMapAttr::get(op.map());
+    // Set the nap attribute
+    source.setAttr(AffineVectorLoadOp::getMapAttrName(), srcMapAttr);
     auto reduce = createVectorReduction(rewriter, op, source.getResult());
-    rewriter.create<AffineVectorStoreOp>(op.getLoc(), ArrayRef<Type>{}, reduce,
-                                         op.mem(), op.idxs());
+    auto dest = rewriter.create<AffineVectorStoreOp>(
+        op.getLoc(), ArrayRef<Type>{}, reduce, op.mem(), op.idxs());
+    // Set the nap attribute
+    auto destMapAttr = AffineMapAttr::get(op.map());
```
Comment (jbruestle) [inline pmlc/conversion/pxa_to_affine/pxa_to_affine.cc:248]: Both src + dest take the same map (and thus the same attribute), no need to call AffineMapAttr::get twice.

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc [hunk 8]
```diff
@@ -236,9 +237,16 @@ struct AffineVectorReduceOpConversion
                ConversionPatternRewriter &rewriter) const override {
     auto source = rewriter.create<AffineVectorLoadOp>(
         op.getLoc(), op.getVectorType(), op.mem(), op.idxs());
+    // Get an attribute form of the map
+    auto srcMapAttr = AffineMapAttr::get(op.map());
+    // Set the nap attribute
```
Comment (jbruestle) [inline pmlc/conversion/pxa_to_affine/pxa_to_affine.cc:242]: Typeo, although a nap attribute sounds really good right now...

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc (hunk 7 above)
Comment (LLITCHEV) [reply pmlc/conversion/pxa_to_affine/pxa_to_affine.cc:248]: Sure! I had it that way, but I have seen LLVM having issues with sharing state objects like this. I choose to stay on the safe side, but I'll change it.

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc (hunk 8 above)
Comment (LLITCHEV) [reply pmlc/conversion/pxa_to_affine/pxa_to_affine.cc]: Fixed

File: pmlc/conversion/pxa_to_affine/pxa_to_affine.cc (hunk 7 above)
Comment (LLITCHEV) [reply pmlc/conversion/pxa_to_affine/pxa_to_affine.cc]: Changed.

Comment (jbruestle) [reply pmlc/conversion/pxa_to_affine/pxa_to_affine.cc]: So attributes + types are interned in MLIR (i.e. they are copied into a big buffer in the MLIR context that is never freed until context destruction) and so they are always safe (and cheap) to reuse aggressively, but constructing them (i.e. interning) is mildly expensive because it involves checking if an identical value exists, and if not, copying and compacting the memory representation across.
